### PR TITLE
Remove Python 3.9 in favour of 3.10 as minimum

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** Remove Py 3.9 from the CI runners and package metadata.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mlxd** [comment]
[sc-65091]

**Comment by mlxd** [comment]
Related works: https://github.com/PennyLaneAI/pennylane-lightning/pull/891 https://github.com/PennyLaneAI/catalyst/pull/1104

**Comment by astralcai** [comment]
Starting Python 3.10 we can use `|` to replace `Union` in type hints, and types from `typing` can directly be used in `isinstance` checks. We can potentially update the code base to take advantage of this.

File: .github/workflows/interface-unit-tests.yml [hunk 1]
```diff
@@ -63,23 +63,23 @@ jobs:
...
-            "jax-tests": ["3.9", "3.12"],
-            "all-interfaces-tests": ["3.9"],
-            "external-libraries-tests": ["3.9"],
-            "qcut-tests": ["3.9"],
-            "qchem-tests": ["3.9"],
-            "gradients-tests": ["3.9"],
-            "data-tests": ["3.9", "3.10"],
-            "device-tests": ["3.9"]
+            "default": ["3.10", "3.11", "3.12"],
+            "torch-tests": ["3.01", "3.12"],
```
Comment (mlxd) [inline .github/workflows/interface-unit-tests.yml:73]:
```suggestion
            "torch-tests": ["3.10", "3.12"],
```

**Comment by mlxd** [comment]
Note: all CI test labels will be updated before merging.

**Review by mudit2812 — APPROVED** [review]
I'm surprised pylint isn't breaking. Maybe worth running CI with the full suite before merging.

**Review by astralcai — APPROVED** [review]

**Comment by mlxd** [comment]
> I'm surprised pylint isn't breaking. Maybe worth running CI with the full suite before merging.

Yep, good call. I'll add the label and retrigger.

**Comment by mlxd** [comment]
> Do we want to pin `python_requires` in `setup.py`?

I think this is fine to leave free for now.

**Comment by mlxd** [comment]
> `python_requires`

Actually, in hindsight I'll set this.

**Pull request merged**